### ISSUE TITLE
Add codecs for singleton types

### ISSUE THREAD

**Issue by alexarchambault** [travisbrown/circe#100]
It would be nice to have codecs for singleton types (``Witness.`"value".`.T`` and the like), to use e.g. like
```scala
case class CC(
  i: Int,
  message: Witness.`"test"`.T
)
```
allowing to force some fields to a given value.

I did that in argonaut-shapeless, and it's working nicely.

Just wanted to enquire opinions, before possibly sending a PR. Also the `generic` module seems the more likely for that, as it depends on shapeless, although this is not fully satisfying: codecs for singleton types are just extra codecs, independent from automatic derivation.

**Comment by travisbrown** [comment]
Are there practical use cases for this? The refined module makes sense to me, but this seems to duplicate the role of other (less clunky) validation mechanisms.

I agree that this doesn't seem to fit nicely in `generic`, and while it'd only take a very simple macro to add it to `core`, I'm not sure we want to give up circe's macro-free status for something without a clear payoff.

**Comment by alexarchambault** [comment]
I typically use these when the underlying protocol expects a constant value. This allows to put / expect constant values in JSON objects, while still using auto-codecs.

**Comment by fthomas** [comment]
If you're going to add support for refined, you could accomplish the same with
```scala
case class CC(
  i: Int,
  message: String @@ Equal[Witness.`"test"`.T]
)
```

**Comment by alexarchambault** [comment]
@fthomas I agree, it's nice to have the simple ``Witness.`"test"`.T`` too.

I don't see it as a validation mechanism... It's just a way to put constants in the output.

These would just be codecs for singleton types, that are around anyway.

**Comment by travisbrown** [comment]
This is resolved by #162, which is now in master.
